Add ReferenceCount for Image to solve image operation competition

moypray opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

Fixed issue: 
  1.  https://github.com/moby/moby/issues/29653
        Not using the solution in https://github.com/moby/moby/pull/32293
  2.  protect build image & rmi operation competition.
  3. protect push & rmi operation competition. (like saving operation).
  4. protect create container & rmi competition.

**- What I did**
Fix image operation competition.
Just lots of issues mentioned, image operation need to pretect. especially build, save, push & rmi operations.

**- How I did it**

Add an `ReferenceCount` for each image, when we need to protect our image, we just use
```
  imageStore.HoldOn(imageID)   // increase  the refcount of the image.
  ... do something ....
  imageStore.HoldOff(imageID)  // decrease the refcount.
```

**- How to verify it**
1.  Saving image follow  https://github.com/moby/moby/issues/29653
2.  write a script for example create container & rmi operation:
    the container may not be startup.
```
#!/bin/bash
docker run -d ubuntu sleep 2000 &
docker rmi ubuntu
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

**- A picture of a cute animal (not mandatory but encouraged)**

moypray commented:
I have collected all the image competition:

Operation | Rmi Image | Solution
------------ | ------------- | -------------
Build | should protect |  RefCount Protection
PULL |  should protect | NOT protected yet
PUSH |  should protect | RefCount protection
Save |  should protect | RefCount protection
Load |  should protect | NOT protected yet
images |   no requirement |  -
inspect |   no requirement |  -
Tag  |   need protection |  need referenceStore and imageStore protection togother
  |  |   
create container   | should protect | RefCount Protection
commit container | should protect | rmi operation will check
export   container | should protect | rmi operation will check


I think most of operations could be protected by RefCount. but load, tag, pull need the referenceStore and imageStore protection together.

moypray commented:
I will rework my PR now, the PRs works fine in my local version, just cherry-pick here.

cpuguy83 commented:
```
05:46:31 --- FAIL: TestRunWithBuildArgs (0.00s)
05:46:31 panic: assignment to entry in nil map [recovered]
05:46:31 	panic: assignment to entry in nil map
05:46:31 
05:46:31 goroutine 49 [running]:
05:46:31 testing.tRunner.func1(0xc42047e270)
05:46:31 	/usr/local/go/src/testing/testing.go:622 +0x29d
05:46:31 panic(0xe637a0, 0xc420480220)
05:46:31 	/usr/local/go/src/runtime/panic.go:489 +0x2cf
05:46:31 github.com/docker/docker/builder/dockerfile.(*Builder).holdOnImage(0xc42045b130, 0x0, 0x0, 0xc420459380, 0x0)
```

tonistiigi commented:
This is being addressed for builder by #32904/(partly in #33054). The other areas(load/save) should use the same method of using layer ref count. `push/pull` shouldn't have an issue afaik. `tag` is bit different case and should be handled in combination with the image delete logic in daemon.

moypray commented:
@tonistiigi   I think we need a `ReferenceCount` in image store.  We can NOT protect image `ONLY` by layer reference. For example:
```
Process 1:
  docker build -t test .   // here will use some cached image.
Process 2:
  docker rmi <ImageCached>   // delete the cached image at the same time.
```

If we only use layer reference to protect image competition just like save operation. the Image structure may be deleted. Error like `unrecognized image ID` will be reported. These days I am working on lots of issues like this (image operation competition).

Take 'save operation'  #32293 example, just use `layerStore.Get` to protect the image, it is not enough too.
Save image operation:
```
1. parseNames()  // call imageStore.Get(id) to get imageID. 
2. use layerStore.Get to protect all the layers.
3. while in saveImage process, it still will call imageStore.Get to get images, 
```
if a deletion operation comes between process 2 and 3 , it does not remove the layers but the image config. the save operation still failed. Error `unrecognized image ID` will be reported too.

Of cause, we could refactory the save operation . but I think `ReferenceCount`  is a better way.
And we could use this to protect more operations like 'creating container' & rmi image operations.

My local docker version is older than this, If the solution is good, I will rework this PR. if not , I will work on the origin design which we talked in #32293

tonistiigi commented:
In #32904 there is nothing using the image ID, only layer is used for executing commands.

For save, the references and config need to be taken before the `saveImages` cycle starts.

moypray commented:
@tonistiigi  I reworked my PR, keep the save and create container protection against the rmi operation.

Do you think the `RefCount` is useful to image??

moypray commented:
The builder code is much different from my local version. I need some time to read. @

tonistiigi commented:
> Do you think the RefCount is useful to image??

No, I think we should stick to layer references. In addition to the maintainence and leaking responsibilities between packages, this has UX issues. Image deletion will randomly start to fail and when image deletion receives an error it ends up in corrupted state where the tags have been removed but not the image itself. It's also racey as image may already be deleted when hold is called.

moypray commented:
@tonistiigi   I will rework  #32293 first.

moypray committed:
Add ReferenceCount for Image to solve image operation competition

Description:
ImageStore do not have an efficiant mechanism to protect the image operation competition.
For example:
  1. One proecess is saving an image.
  2. Another process delete this image.
Then will failed to save the image.

Solution:
  So I have added an image ReferenceCount to ImageStore. If someone need
  to protect an image, he may call:

```
   imageStore.HoldOn(ImageID)
   ... do something ...
   imageStore.HoldOff(ImageID)
```

This PR also fixed two issue:
 1. Creating Container & rmi the image competition.
 2. Saving image & rmi the image comptition.

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

cpuguy83 commented:
What's the status of this one?

moypray commented:
@cpuguy83   Tonistiigi  think the solution is a not good one.

So I do not work on it later. If the solution is OK, I could rework it. If not, we could close it now.

tonistiigi commented:
Load/save part of this was merged in #32293 . I'm not sure I know the push/pull problems but they could probably be fixed similarly to load/save. The build is being worked on separately(mostly merged already). Tagging definitely still has a race, but that is more related to the image delete logic in the daemon and not image store.

moypray commented:
That's fine, Close this one.
